mongo/integration/gridfs_spec_test.go

```diff
@@ -0,0 +1,498 @@
+	gfsFiles          = "fs.files"
+	gfsChunks         = "fs.chunks"
+	gfsExpectedFiles  = "expected.files"
+	gfsExpectedChunks = "expected.chunks"
+)
+
+var (
+	gfsDeadline    = time.Now().Add(time.Hour)
+	downloadBuffer = make([]byte, 100)
+)
+
+func TestGridFSSpec(t *testing.T) {
+	mt := mtest.New(t, noClientOpts)
+	defer mt.Close()
+
+	for _, file := range jsonFilesInDir(mt, gfsTestsDir) {
+		mt.Run(file, func(mt *mtest.T) {
+			runGfsTestFile(mt, path.Join(gfsTestsDir, file))
+		})
+	}
+}
+
+func runGfsTestFile(mt *mtest.T, filePath string) {
+	content, err := ioutil.ReadFile(filePath)
+	assert.Nil(mt, err, "ReadFile error for %v: %v", filePath, err)
+
+	var testFile gfsTestFile
+	err = bson.UnmarshalExtJSONWithRegistry(specTestRegistry, content, false, &testFile)
+	assert.Nil(mt, err, "UnmarshalExtJSONWithRegistry error: %v", err)
+
+	for _, test := range testFile.Tests {
+		mt.Run(test.Description, func(mt *mtest.T) {
+			runGfsTest(mt, test, testFile)
+		})
+	}
+}
+
+func runGfsTest(mt *mtest.T, test gfsTest, testFile gfsTestFile) {
+	chunkSize := setupGfsTest(mt, testFile.Data)
+	if chunkSize == 0 {
+		chunkSize = gridfs.DefaultChunkSize
+	}
+	bucket, err := gridfs.NewBucket(mt.DB, options.GridFSBucket().SetChunkSizeBytes(chunkSize))
+	assert.Nil(mt, err, "NewBucket error: %v", err)
+	err = bucket.SetWriteDeadline(gfsDeadline)
+	assert.Nil(mt, err, "SetWriteDeadline error: %v", err)
+	err = bucket.SetReadDeadline(gfsDeadline)
+	assert.Nil(mt, err, "SetReadDeadline error: %v", err)
+
+	arrangeGfsCollections(mt, test.Arrange)
+	switch test.Act.Operation {
+	case "upload":
+		executeGfsUpload(mt, test, bucket)
+		checkGfsResults(mt, test)
+		clearGfsCollections(mt)
+
+		arrangeGfsCollections(mt, test.Arrange)
+		executeGfsUploadFromStream(mt, test, bucket)
+		checkGfsResults(mt, test)
+	case "download":
+		executeGfsDownload(mt, test, bucket)
+		checkGfsResults(mt, test)
+
+		executeGfsDownloadToStream(mt, test, bucket)
+		checkGfsResults(mt, test)
+	case "download_by_name":
+		executeGfsDownloadByName(mt, test, bucket)
+		checkGfsResults(mt, test)
+
+		executeGfsDownloadByNameToStream(mt, test, bucket)
+		checkGfsResults(mt, test)
+	case "delete":
+		executeGfsDelete(mt, test, bucket)
+		checkGfsResults(mt, test)
+	}
+}
+
+func checkGfsResults(mt *mtest.T, test gfsTest) {
+	if test.Assert.Error != "" {
+		// don't compare collections in non-error cases
+		return
+	}
+	compareGfsCollections(mt, gfsExpectedChunks, gfsChunks)
+	compareGfsCollections(mt, gfsExpectedFiles, gfsFiles)
+}
+
+func compareGfsCollections(mt *mtest.T, expected, actual string) {
+	expectedCursor, err := mt.DB.Collection(expected).Find(mtest.Background, bson.D{})
+	assert.Nil(mt, err, "Find error for collection %v: %v", expected, err)
+	actualCursor, err := mt.DB.Collection(actual).Find(mtest.Background, bson.D{})
+	assert.Nil(mt, err, "Find error for collection %v: %v", actual, err)
+
+	var idx int
+	for expectedCursor.Next(mtest.Background) {
```

Comment:
Let's also assert that actualCursor is exhausted after expectedCursor is, i.e. `assertFalse(actualCursor.Next())` at the end of the loop

Reply:
Added.